### PR TITLE
Fix npc mounting, add permanent skins to player npcs

### PR DESCRIPTION
**Pull request by sanjay900**
https://github.com/CitizensDev/Citizens2/issues/189

### CODE REVIEW

**Comment by sanjay900** [comment]
Permanent skins work by using ./npc skin -b playername. When you do this, the npc permanently stores the skin blob for this skin, and that means that even if the player that you have this skin from changes their skin, your npc's skin won't change since it is using a previous skin blob.

**Comment by sanjay900** [comment]
Oh yeah, and a skin set with -b will load instantly since it doesnt query mojangs api for the skin since it already has it.

**Comment by mcmonkey4eva** [comment]
... is it -b or -p?

**Comment by sanjay900** [comment]
-p
Sorry, i changed it half way through because it made more sense.

**Comment by mcmonkey4eva** [comment]
Also, this needs longer-term testing.

Last I heard, skin blobs used outside NPC heads expire after ~1 day.

**Comment by sanjay900** [comment]
Ive had on on for the last few weeks....

**Comment by sanjay900** [comment]
So clearly they last for a pretty long time

**Comment by mcmonkey4eva** [comment]
Also, shouldn't this be two separate PR's, as it covers two separate things?

-- ... oh, really? 0.o did something in Mojang's... I don't....

-- Also, make it say -p in the help info if it's -p

File: src/main/java/net/citizensnpcs/commands/NPCCommands.java [hunk 1]
```diff
@@ -1229,11 +1229,12 @@ public void skeletonType(CommandContext args, CommandSender sender, NPC npc) thr
 
     @Command(
             aliases = { "npc" },
-            usage = "skin (-c) [name]",
-            desc = "Sets an NPC's skin name",
+            usage = "skin (-c,b) [name]",
```
Comment (mcmonkey4eva) [inline src/main/java/net/citizensnpcs/commands/NPCCommands.java]: Ehm.

**Comment by sanjay900** [comment]
The whole snapshot system is actually something that mojang did on purpose, since thats how Player heads were sussposed to work

**Comment by sanjay900** [comment]
So they should last forever, since mojang did it on purpose because of decorative player heads.

**Comment by sanjay900** [comment]
And i was going to make it two seperate PRs but then i pushed it and it automatically added it to this one so yeah...

**Comment by sanjay900** [comment]
Also, funnily enough this actually makes the -c flag that existed before work, because flags= was missing before in the @command annotation

File: src/main/java/net/citizensnpcs/commands/NPCCommands.java (hunk 1 above)
Comment (mcmonkey4eva) [inline src/main/java/net/citizensnpcs/commands/NPCCommands.java]: ^

**Comment by mcmonkey4eva** [comment]
I don't think the flags= is actually /required/, as -c was known to be working.

**Comment by sanjay900** [comment]
Weird, i tested it before and it didnt work which was weird.

**Comment by sanjay900** [comment]
I might do a bit more long term testing though. Since i haven't tested this specific code.

**Comment by sanjay900** [comment]
Just looking through the client though, it doesn't appear to check the timestamp of a skin. And i know the skins do get sent to the client because i have a custom tab list that has a signed skin blob on it thats been there for a few weeks. Ill keep a npc with a skin blob on it sitting there for a while and we will see what happenes.

**Comment by fullwall** [comment]
You have to be careful because the signature may change... I think this needs more testing.

**Comment by sanjay900** [comment]
Yeah. Ill keep an npc there with a skin blob on it for a while. Its survived longer than a day though, so thats nice.

**Comment by sanjay900** [comment]
The signatures though... Ive had several that still work for a few weeks, so i don't know if thats long enough for testing. Looking through the client code, The tablist and the player use the same code to fetch a players skin, and although ive only done testing on the tablist, it seems the signature stays valid for quite a while.

**Comment by Roos-Skywalker** [comment]
Permanent skins absolutely sounds like a fabulous idea.

**Comment by sanjay900** [comment]
ive had one skin survive since  Feb 3, 2015, so thats always nice. It seems that it may need a bit more editing, because ocassionally i have to respawn an npc for its skin to show.

**Comment by tronman2020** [comment]
The -p and -c flags are unrecognized by citizens

**Comment by Morphan1** [comment]
@sanjay900 Have you found any limit to signature validity yet, or made any progress with testing?

**Comment by sanjay900** [comment]
It appears to keep working for a while, though somethings up with npcs.
Testing on real players, skins stay valid for a long time. With npcs its
the same although sometimes the skin isnt sent.

**Comment by mcmonkey4eva** [comment]
Yeah skin not sending is a glitch in main C2, not your fault.

**Comment by sanjay900** [comment]
Oh, okay. Besides that, I know for a fact that skin signature timestamps
arent checked, so permanent skins should work pretty well.

**Comment by sanjay900** [comment]
If skins are fixed, This should work fine now.

**Comment by Roos-Skywalker** [comment]
Very well but how do I use this? I know you are a kind person Sanjay900 but please treat me as if I am retarded child who doesn't know a thing about it, because I really don't. I know how to set up a proper server but that's it really.

**Comment by sanjay900** [comment]
Wait for it to get merged in. Right now, it isnt really updated to the
latest version of Citizens anyway, so ill have to work on that soon.

**Comment by mcmonkey4eva** [comment]
@fullwall This PR looks pretty solid still

**Pull request merged**